lib/active_merchant/billing/gateways/visanet_peru.rb

```diff
@@ -0,0 +1,239 @@
+      # end
+
+      def verify(credit_card, options={})
+        MultiResponse.run(:use_first_response) do |r|
+          r.process { authorize(100, credit_card, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
+
+      # def store(payment_method, options = {})
+      #   post = {}
+      #   add_payment_method(post, payment_method)
+      #   add_customer_data(post, options)
+
+      #   commit("store", post)
+      # end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        #JSON.
+        transcript.
+          gsub(%r((Authorization: Basic )\w+), '\1[FILTERED]').
+          gsub(%r((\"cardNumber\\\":\\\")\d+), '\1[FILTERED]').
+          gsub(%r((\"cvv2Code\\\":)\d+), '\1[FILTERED]')
+      end
+
+      private
+
+      CURRENCY_CODES = Hash.new{|h,k| raise ArgumentError.new("Unsupported currency: #{k}")}
+      CURRENCY_CODES["USD"] = 840
+      CURRENCY_CODES["PEN"] = 604
+
+      def add_invoice(post, money, options)
+        post[:amount] = amount(money).to_f
+        # Visanet Peru expects a 9-digit numeric purchaseNumber
+        post[:purchaseNumber] = options[:order_id]
+        post[:externalTransactionId] = options[:order_id]
+        post[:currencyId] = CURRENCY_CODES[options[:currency] || currency(money)]
+      end
+
+      def add_payment_method(post, payment_method)
+        post[:firstName] = payment_method.first_name
+        post[:lastName] = payment_method.last_name
+        post[:cardNumber] = payment_method.number
+        post[:cvv2Code] = Integer(payment_method.verification_value, 10)
+        post[:expirationYear] = format(payment_method.year, :four_digits)
+        post[:expirationMonth] = format(payment_method.month, :two_digits)
+      end
+
+      def add_customer_data(post, options)
+        post[:email] = options[:email]
+        antifraud = {}
+        billing_address = options[:billing_address] || options[:address]
+        if (billing_address)
+          antifraud[:billTo_street1] = billing_address[:address1]
+          antifraud[:billTo_city] = billing_address[:city]
+          antifraud[:billTo_state] = billing_address[:state]
+          antifraud[:billTo_country] = billing_address[:country]
+          antifraud[:billTo_postalCode]    = billing_address[:zip]
+        end
+        post[:antifraud] = antifraud
+      end
+
+      def add_device_fingerprint_data(post, options)
+        post[:antifraud][:deviceFingerprintId] = options[:device_fingerprint_id]
+      end
+
+      def add_merchant_define_data(post, options)
+        if (merchantDefineData = options[:merchant_define_data])
+          post[:antifraud][:merchantDefineData] = merchantDefineData
+        end
+      end
+
+      def commit(action, params, options)
+        case action
+        when "authorize"
+          url = base_url() + "/" + options[:merchant_id]
+          method = :post
+        when "capture"
```

Comment:
For these three `purchaseNumber` based operations, why don't we just pass in the right string to use in the URL vs. having to map from a "capture" to a "deposit"? So the calls to `commit` will look like:

```ruby
commit("deposit", post, options)
commit("void", post, options)
commit("cancelDeposit", post, options)
```

I guess you can leave "authorize" since it has no direct mapping.